lib/IRGen/GenBuiltin.cpp

```diff
@@ -1537,3 +1539,682 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+    
+    IGF.Builder.emitBlock(endBB);
+  }
+
+public:
+  void assignWithCopy(IRGenFunction &IGF, Address dest, Address src,
+                      SILType T, bool isOutlined) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddrLoop(IGF, T,
+                        [&](ArrayRef<Address> destAndSrc) {
+                          Element.assignWithCopy(IGF, destAndSrc[0],
+                                                 destAndSrc[1],
+                                                 eltTy, isOutlined);
+                        }, {dest, src});
+  }
+
+  void assignWithTake(IRGenFunction &IGF, Address dest, Address src,
+                      SILType T, bool isOutlined) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddrLoop(IGF, T,
+                        [&](ArrayRef<Address> destAndSrc) {
+                          Element.assignWithTake(IGF, destAndSrc[0],
+                                                 destAndSrc[1],
+                                                 eltTy, isOutlined);
+                        }, {dest, src});
+  }
+
+  void initializeWithCopy(IRGenFunction &IGF, Address dest, Address src,
+                          SILType T, bool isOutlined) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddrLoop(IGF, T,
+                        [&](ArrayRef<Address> destAndSrc) {
+                          Element.initializeWithCopy(IGF, destAndSrc[0],
+                                                     destAndSrc[1],
+                                                     eltTy, isOutlined);
+                        }, {dest, src});
+  }
+  
+
+  void initializeWithTake(IRGenFunction &IGF, Address dest, Address src,
+                          SILType T,
+                          bool isOutlined,
+                          bool zeroizeIfSensitive) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddrLoop(IGF, T,
+                        [&](ArrayRef<Address> destAndSrc) {
+                          Element.initializeWithTake(IGF, destAndSrc[0],
+                                                     destAndSrc[1],
+                                                     eltTy, isOutlined,
+                                                     zeroizeIfSensitive);
+                        }, {dest, src});
+  }
+  
+  virtual void destroy(IRGenFunction &IGF, Address address, SILType T,
+                       bool isOutlined) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddrLoop(IGF, T,
+                        [&](ArrayRef<Address> elt) {
+                          Element.destroy(IGF, elt[0], eltTy, isOutlined);
+                        }, {address});
+  }
+};
+
+template<typename BaseTypeInfo>
+class FixedArrayTypeInfoBase : public ArrayTypeInfoBase<BaseTypeInfo> {
+protected:
+  using ArrayTypeInfoBase<BaseTypeInfo>::Element;
+  const unsigned ArraySize;
```

Comment:
This has been zero-bounded?

Reply:
Yes. Zero or negative sizes get an `EmptyTypeInfo`.